devicemapper: remove container rootfs mountPath after umount

cyphar opened the pull request:
libdm currently has a fairly substantial DoS bug that makes certain
operations fail on a libdm device if the device has active references
through mountpoints. This is a significant problem with the advent of
mount namespaces and MS_PRIVATE, and can cause certain --volume mounts
to cause libdm to no longer be able to remove containers:

    % docker run -d --name testA busybox top
    % docker run -d --name testB -v /var/lib/docker:/docker busybox top
    % docker rm -f testA
    [fails on libdm with dm_task_run errors.]

This also solves the problem of unprivileged users being able to DoS
docker by using unprivileged mount namespaces to preseve mounts that
Docker has dropped.

[![Walter](https://farm5.staticflickr.com/4171/34523002411_5910bad68a_k.jpg)](https://flic.kr/p/UAFpRv)

###### [Walter](https://flic.kr/p/UAFpRv) by [Alexander Lemke](https://www.flickr.com/photos/147569256@N08/).

Closes #34542
Signed-off-by: Aleksa Sarai <asarai@suse.de>

/cc @rhvgoyal This is an alternative to #34542 
/cc @vrothberg

cyphar committed:
devicemapper: remove container rootfs mountPath after umount

libdm currently has a fairly substantial DoS bug that makes certain
operations fail on a libdm device if the device has active references
through mountpoints. This is a significant problem with the advent of
mount namespaces and MS_PRIVATE, and can cause certain --volume mounts
to cause libdm to no longer be able to remove containers:

  % docker run -d --name testA busybox top
  % docker run -d --name testB -v /var/lib/docker:/docker busybox top
  % docker rm -f testA
  [fails on libdm with dm_task_run errors.]

This also solves the problem of unprivileged users being able to DoS
docker by using unprivileged mount namespaces to preseve mounts that
Docker has dropped.

Signed-off-by: Aleksa Sarai <asarai@suse.de>

cyphar commented:
It's a bit of a pain that Jenkins doesn't test any graphdrivers other than AUFS and (presumably) overlay for non-Ubuntu machines.

dnephin commented:
We should have separate suites for testing the graph drivers, so we can run the suite for any supported drivers.

dnephin commented on integration/util/kernel.go:
```diff
@@ -0,0 +1,18 @@
+//   -1 if lval < rval
+//    0 if lval = rval
+//    1 if lval > rval
+func MustKernelCmp(rval kernel.VersionInfo) int {
```
Could this be inlined into `run_test.go`? Do you expect to use it outside of container tests?

dnephin commented on integration/container/run_test.go:
```diff
@@ -0,0 +1,65 @@
+	// tested ocurred). We run it anyway just to be safe.
+	driver, _ := daemonStorageDriver(context.Background(), client)
+	if driver != "devicemapper" {
+		t.Logf("Active driver (%s) is not devicemapper, continuing anyway.", driver)
```
I think this should `t.Skip()` if it's not devicemapper.

cyphar replied:
I think it would be a good measure to make sure the bug doesn't happen on other storage drivers too. I should probably just remove this check all-together.

dnephin replied:
I don't think this is the right way to test storage drivers.  These "integration" tests are very slow, we need to be conservative about the test we add here (see [TESTING](https://github.com/moby/moby/blob/master/TESTING.md)).

We should be able to test storage drivers without the daemon running, which makes it much easier and faster to test them. It looks like each graphdriver has a test suite which uses `graphtest`. We should test this case in that suite.

Why is it not possible to just check that the mount point doesn't exist anymore from that test suite?

cyphar replied:
I would need to check again, but it's not that it's impossible, it's that I'd have to effectively start re-implementing parts of `runc` inside the test framework to make sure we're testing the right interaction between mount namespaces and `libdm`.

Just checking if the mountpoint has gone isn't good enough, you need to check that the mount has disappeared in other namespaces.

dnephin commented on integration/container/run_test.go:
```diff
@@ -0,0 +1,65 @@
+	// torvalds/linux@8ed936b5671bfb33d89bc60bdcc7cf0470ba52fe was merged).
+	oldKernel := util.MustKernelCmp(kernel.VersionInfo{Kernel: 3, Major: 18, Minor: 0}) < 0
+	if oldKernel {
+		t.Logf("NOTE: This test may fail on your <3.18 kernels.")
```
Again, probably `t.Skip()`

kolyshkin replied:
You can use `kernel.CheckKernelVersion` from `github.com/docker/docker/pkg/parsers/kernel` here. For example:

```go
import "github.com/docker/docker/pkg/parsers/kernel"
...
oldKernel := !kernel.CheckKernelVersion(3, 18, 0)
...
```

cyphar commented on integration/container/run_test.go:
```diff
@@ -0,0 +1,65 @@
+	// torvalds/linux@8ed936b5671bfb33d89bc60bdcc7cf0470ba52fe was merged).
+	oldKernel := util.MustKernelCmp(kernel.VersionInfo{Kernel: 3, Major: 18, Minor: 0}) < 0
+	if oldKernel {
+		t.Logf("NOTE: This test may fail on your <3.18 kernels.")
```
Some distributions may have backported the fix, so the kernel version alone isn't enough to know whether the test will work (but we don't hard fail on older kernels).

cyphar commented:
@dnephin So where does that leave this PR's tests? Should I not include tests at all (these tests actually currently fail silently because the error handling in the new `integration` scheme doesn't exist yet -- and we'll eventually need to recreate the old `integration-cli` testing setup), or should I start working on making it possible to have `devicemapper` integration tests?

dnephin commented:
> these tests actually currently fail silently because the error handling in the new integration scheme doesn't exist yet

I don't understand this. What do you mean by error handling doesn't exist?

> while integration doesn't have any significant infrastructure for testing

It's a work in progress. If you see something significant that is missing, please let me know.

> not to mention that integration-cli has three orders of magnitude more tests that we can't touch anymore because of the deprecation warnings

You can change tests, you just can't add new ones

> So where does that leave this PR's tests? 

The ideal place for this test would be in `daemon/graphdriver/devmapper/devmapper_test.go`.  I don't understand yet why it's not possible to exercise the bug there. If it's not possible I think what you have now is the right idea, but we should skip it when the graphdriver is not devicemapper

rhvgoyal commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2416,6 +2417,15 @@ func (devices *DeviceSet) UnmountDevice(hash, mountPath string) error {
+	if err := system.EnsureRemoveAll(mountPath); err != nil {
+		logrus.Debugf("devmapper: error doing a removeall on unmounted device %s: %v", mountPath, err)
+	}
+
```
Why use EnsureRemoveAll? This function is trying to do so much that I don't even understand what it is doing. After lazy unmount, why don't we simply do a `os.Remove()` ?

So idea is that use removing directory to yank the mount points which might have leaked into other mount namespaces and this will only work on newer kernels. I think this should work reasonably well. Little concerned about some use case being broken somewhere. There have been quite a few changes w.r.t referece counting. Especially make sure that the case of docker daemon being shutdown/start works while containers were running the whole time.

cyphar replied:
Yeah, `os.Remove` is probably a better choice.

> Especially make sure that the case of docker daemon being shutdown/start works while containers were running the whole time.

Can you explain what you mean here? Are you referring to the "live restore" stuff?

rhvgoyal replied:
Yes, I was referring to live restore functionality.

rhvgoyal commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2416,6 +2417,15 @@ func (devices *DeviceSet) UnmountDevice(hash, mountPath string) error {
 	}
 	logrus.Debug("devmapper: Unmount done")
 
+	// Remove the mountpoint here. This is necessary to avoid cases where a
```
It might be good to tell that removing directory will remove the mount point on this directory from all the mount namespaces.

rhvgoyal commented:
@cyphar We will require similar change in Remove() path as well? That is, make sure device is unmounted, then remove mnt directory and then try device deactivation and deletion?

rhvgoyal commented:
cc @rhatdan

cyphar commented:
@rhvgoyal Yes probably. I'll need to figure out what's the best place to put this code (I'd prefer to not have to duplicate it in both paths).

cyphar commented:
Actually @rhvgoyal I don't think we need this in the Remove path. The Remove path is only called from the driver after it's been `Put` and removed (and `Put` calls `UnmountDevice` which removes the mountpoint). I'm testing the live restore stuff now.

cyphar commented:
It looks like our linter is giving errors about the standard library:

```
01:47:56 ../../../../../usr/local/go/src/os/user/lookup.go:38:9:warning: undeclared name: listGroups (unconvert)
01:47:56 ../../../../../usr/local/go/src/os/user/lookup.go:9:9:warning: undeclared name: current (unconvert)
01:47:56 ../../../../../usr/local/go/src/os/user/lookup.go:15:9:warning: undeclared name: lookupUser (unconvert)
01:47:56 ../../../../../usr/local/go/src/os/user/lookup.go:21:9:warning: undeclared name: lookupUserId (unconvert)
01:47:56 ../../../../../usr/local/go/src/os/user/lookup.go:27:9:warning: undeclared name: lookupGroup (unconvert)
01:47:56 ../../../../../usr/local/go/src/os/user/lookup.go:33:9:warning: undeclared name: lookupGroupId (unconvert)
```

cyphar commented:
@rhvgoyal I've tested `--live-restore` quite a bit and it looks like it doesn't affect anything. It looks like Docker doesn't do any graphdriver cleanups (including `Put`) if you have `--live-restore`. That makes sense because otherwise `docker` would hit a bunch of errors when you're using `--live-restore`.

cyphar commented:
@dnephin Because of the weird errors that look like the linters aren't doing the right thing, I've just opted to not add a test for it. I would've preferred being able to add a test, but we can follow up with my test after the lint issue is resolved.

dnephin commented:
What lint errors?

cyphar commented:
@dnephin [These ones.](#issuecomment-326869266) You can't see them anymore because I dropped the test commit.

dnephin commented:
hmm, not sure what that's about. What did you change to remove them?  

If you have a commit that can reproduce can you push it here , or link me to it?

cyphar commented:
@dnephin Here's a branch with the top commit being the one that causes the error. I fixed the issue by just dropping that commit, I'm not sure which part of that commit was causing the issue. https://github.com/cyphar/docker/tree/docker-lint-error

rhvgoyal commented:
@cyphar This patch LGTM. 

If need be, we can make similar changes in Remove() path later.

dnephin commented:
@cyphar I've seen those lint errors happen on occasionally on my branch as well. I think it's a weird error with jenkins, but doesn't have anything to do with your test code. If you build it again they should be gone.

thaJeztah commented:
ping @cpuguy83 @dnephin what's the status on this one?

dnephin commented:
I think this just needs a unit test

cyphar commented:
@dnephin I'll see if I can work up a sane unit test next week (I'm at OSS this week).

vieux commented:
@kir can you please take a look at this one ?

trapier commented:
On a kernel with torvalds/linux@8ed936b5671bfb33d89bc60bdcc7cf0470ba52fe and 

```
  "storage-driver": "devicemapper",
   "storage-opts": [
     "dm.thinpooldev=/dev/mapper/docker-thinpool",
     "dm.use_deferred_removal=true",
     "dm.use_deferred_deletion=true"
   ]
```

... the test case in the PR description does not fail.  Is the goal to remove the need to set `"dm.use_deferred_deletion=true"`?

rhvgoyal commented:
@trapier, If we can remove leaked mount points, then deferred deletion will not even kick in. For now I will not be worried about it being enabled by default. If it causes trouble, we can think about it.

rhvgoyal commented:
@cyphar BTW, will we need similar fix for overlay2 as well? I am assuming that if mount point leaks, then container removal will fail in overlay2 as well.

cyphar commented:
@rhvgoyal I don't think overlay will have this problem. The problem with devicemapper stems from the problem that you cannot do a cleanup of the device if there is a live mountpoint. Overlay does not suffer from this problem (since the cleanup is just the removal of directories which is the same trick as we do here), but I can double check this if you like.

@trapier Older distributions don't have deferred removal, due to old LVM versions. Also I've seen some distributions where that option doesn't actually do what it says on the tin (though to be fair that's a distribution bug).

rhvgoyal commented:
@cyphar ok, got it. You are right. I quickly tested docker-1.13 and it does not seem to show the problem. Thanks.

cyphar commented:
/ping? Is there anything left? I started working on a `devicemapper` unit test, but found it would not be reasonably doable -- so we can just do without it.

rhvgoyal commented:
I am fine with the patch. I guess docker maintainers have to provide thumbs up.

yongtang commented:
/cc @cpuguy83 please take a look.

cpuguy83 reviewed (approved):
LGTM

cpuguy83 commented:
ping @kolyshkin

kolyshkin commented:
LGTM, aside from the lack of unit test.

vieux commented:
@cyphar any way to add a test ?

cyphar commented:
@vieux The only way I could see to add a test would be to try to emulate what happens when you create a container's rootfs and then remove it. But in order to test the actual regression you'd need to create a container with `MS_PRIVATE` rootfs propagation that includes the mount (something that I don't feel is a good idea to write in Go given the historical issues with `unshare(2)` and Go).

I used to have an integration test for this, but the `integration-cli` suite has been deprecated, so I'd have to use the now-new-again `integration` suite which proved too tiresome to reinvent the wheel as well as not being clear whether it actually is testing what we care about (since I can't force the test to run on `devicemapper`).

cyphar commented:
*nudge*

cpuguy83 commented:
Maybe a unit test that does uses unshare(1), and can skip if it's not available. Not exactly ideal, but ...

cyphar commented:
The problem with such a test is that `unshare` would possibly break other unit tests. I'll just write a test to make sure that the mountpoint was deleted, I guess...

[Again, this would not be a problem if I could just write a normal integration test... 😞]

dnephin commented:
Why would it break other unit tests?

cyphar commented:
@dnephin The tests all run in one binary invocation, so it could affect other tests (though we could do a `defer unix.Setns`). But in order to test this properly you'd need to actually fork -- which is very painful in Go to do properly (especially in the context of a `go test` run). And there's also the fact that `unix.Unshare` and `unix.Setns` generally aren't to be trusted in Go programs.

cyphar committed:
devmapper: add a test for mount leak workaround

In order to avoid reverting our fix for mount leakage in devicemapper,
add a test which checks that devicemapper's Get() and Put() cycle can
survive having a command running in an rprivate mount propagation setup
in-between. While this is quite rudimentary, it should be sufficient.

We have to skip this test for pre-3.18 kernels.

Signed-off-by: Aleksa Sarai <asarai@suse.de>

cyphar commented:
@dnephin I've added a unit test in `graphdriver/devmapper/...`.

/cc @vieux @cpuguy83

cyphar commented:
`ppc` failure is not related to my change, it looks like some test wrapper broke in the middle of running the suite.

thaJeztah commented:
Restarted PowerPC; I was told there were some connection issue with Jenkins and the PowerPC nodes

cpuguy83 reviewed (approved):
LGTM

Thanks @cyphar!

cyphar commented:
/cc @vdemeester

vdemeester reviewed (approved):
LGTM 🐮

dnephin reviewed (commented):
Thanks for writing the unit test

cyphar commented:
@dnephin Sorry, I never answered your question. I was worried about using `Unshare` in the context of a Go program (historically that's caused nothing but trouble, and I'd argue that it's not solveable in Go because it gives you so little information about the threads that your program runs on). But in the final patch state, I just ended up using `os/exec` which has support for specifying `Unshare` and `Clone` flags that get applied correctly.
